### PR TITLE
QR Code Scanner

### LINKED ISSUE #686: Webcam support for scanning QR codes
Moving funds between phone and MetaMask is cumbersome. 
It would be conveinient to be able scan QR code from phone directly

### PR DESCRIPTION
**Pull request by brunobar79**
Fixes #686 

@alextsg  @danjm  @whymarrh  Whenever any of you have a chance, please take a look at this PR.

Please check out  [this comment](https://github.com/MetaMask/metamask-extension/pull/4897#issuecomment-409394331) before looking at the code to understand the motivation behind certain parts of the code and the different scenarios that needs to be supported.

Thanks!

### CODE REVIEW

**Comment by Maayu2415** [comment]
Stay in add-on.

ในวันที่ ศ. 27 ก.ค. 2018 08:12 Bruno Barbieri <notifications@github.com>
เขียนว่า:

**Comment by brunobar79** [comment]
Here you can see all the different scenarios I'm handling:

1) User coming from the popup (no permission granted yet):
Because you can't ask for permissions while on the popup, I'm forcing fullscreen mode with the modal open automatically and the permission prompt on top.

cuv6cd4vau

2) User coming from the popup (permission previously granted):
lfslre8gdd


3)  User coming from full screen mode (no permission granted yet):
No need to include gif since it's the same as number 1, after it goes to full screen mode

4) User coming from full screen mode (permission previously granted):
No need to include gif since it's the same as number 2, except it's on fullscreen mode

5) User declines permission (and later allows it):
tw9vsukuqq

6) No webcam detected
screen shot 2018-07-31 at 6 32 34 pm

7) Unknown error while trying to access the camera.
screen shot 2018-07-31 at 6 33 42 pm

@cjeria  @bdresser  Please take a look at the screens posted above and send any UX / UI  that you guys have so we can finish this feature.

(I put no effort on 6 and 7 since you might wanna reuse something else that we already have)

Thanks!

**Comment by cjeria** [comment]
@brunobar79 @bdresser What do you guys think about moving the QR code off to the right (see gif)? 

 
Let's introduce a close (X) button to the modal
image

**Comment by bdresser** [comment]
@cjeria makes sense to me 👍 
could do a slightly lighter gray but nbd

**Comment by cjeria** [comment]
We should be careful with colors and make sure we are following web accessibility standards for the color blind. Looks like #f1f1f1 is the lighted per the us gov standards. https://designsystem.digital.gov/components/colors/

**Comment by cjeria** [comment]
So let's make the QR button gray on hover and add a tool tip(?) for additional context since desktop webcam access to scan QR codes is a new-ish blockchain-y ux pattern.

File: ui/app/components/modals/qr-scanner/qr-scanner.component.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        <div className="qr-scanner__image">
+          <img src={'images/webcam.svg'} width={70} height={70} />
+        </div>
+        <div className="qr-scanner__title">
+          { title }
+        </div>
+        <div className={'qr-scanner__error'}>
+          {msg}
+        </div>
+        <div className={'qr-scanner__footer'}>
```
Comment (danjm) [inline ui/app/components/modals/qr-scanner/qr-scanner.component.js]: A small issue that should be fixed before the PR is merged: `CANCEL` and `TRY AGAIN` should be translated.

A slightly bigger issue that could be fixed after this PR is merged: code duplication between this `<div className={'qr-scanner__footer'}>` and `metamask-extension/ui/app/components/page-container/page-container-footer/page-container-footer.component.js`. This need not be a blocker and is really beyond the scope of this PR, but we should create a github issue (or PR) immediately after this to the needed change isn't forgotten.

File: ui/app/components/modals/qr-scanner/qr-scanner.component.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        msg = this.context.t('noWebcamFound')
+      } else {
+        title = this.context.t('unknownCameraErrorTitle')
+        msg = this.context.t('unknownCameraError')
+      }
+    }
+
+    return (
+      <div className="qr-scanner">
+        <div className="qr-scanner__close" onClick={this.stopAndClose}></div>
```
Comment (danjm) [inline ui/app/components/modals/qr-scanner/qr-scanner.component.js]: Not at all an issue with this PR, but we should not have to implement one of these closing 'X's for every modal. I have created a github issue to improve the modal component to provide this feature: https://github.com/MetaMask/metamask-extension/issues/4956

File: ui/app/actions.js [hunk 3]
```diff
@@ -1806,6 +1832,12 @@ function hideAlert () {
   }
 }
 
+function qrCodeDetected (qrCodeData) {
```
Comment (danjm) [inline ui/app/actions.js]: This is a pretty small, nit-picky request, but could you make it explicit what properties are in qrCodeData? In the absence of both tests and formal documentation for this file, make properties explicit makes the code a bit more self documenting. See, for example, the `showQrView`, `updateSendTo` or `setCurrentCurrency` functions in this file.

In this case I am thinking of some variation of:
```
function qrCodeDetected ({ type, values }) {
  return {
    type: actions.QR_CODE_DETECTED,
    value: { type, values },
  }
```

**Review by danjm — CHANGES_REQUESTED** [review]
Only a couple very minor blocking requests. Otherwise, my comments relate to improvements that can be made separate to this PR.

Overall, really great PR. Very solid and smooth integration with the existing codebase.

I am not able to properly QA as I don't have a working smartphone right now.

File: ui/app/components/modals/qr-scanner/qr-scanner.component.js (hunk 2 above)
Comment (brunobar79) [reply ui/app/components/modals/qr-scanner/qr-scanner.component.js]: 100% agree with you and thank you for creating the issue!

File: ui/app/actions.js (hunk 3 above)
Comment (whymarrh) [reply ui/app/actions.js]: Another option might be to add a docblock to the function

Comment (brunobar79) [reply ui/app/actions.js]: Ended up following @whymarrh's suggestion. Thanks!

**Review by danjm — APPROVED** [review]

**Pull request merged**